python/cuvs/cuvs/neighbors/ivf_pq/ivf_pq.pyx

```diff
@@ -0,0 +1,430 @@
+        self.params.force_random_rotation = force_random_rotation
+        self.params.add_data_on_build = add_data_on_build
+        self.params.conservative_memory_allocation = \
+            conservative_memory_allocation
+
+    @property
+    def metric(self):
+        return self._metric
+
+    @property
+    def metric_arg(self):
+        return self.params.metric_arg
+
+    @property
+    def add_data_on_build(self):
+        return self.params.add_data_on_build
+
+    @property
+    def n_lists(self):
+        return self.params.n_lists
+
+    @property
+    def kmeans_n_iters(self):
+        return self.params.kmeans_n_iters
+
+    @property
+    def kmeans_trainset_fraction(self):
+        return self.params.kmeans_trainset_fraction
+
+    @property
+    def pq_bits(self):
+        return self.params.pq_bits
+
+    @property
+    def pq_dim(self):
+        return self.params.pq_dim
+
+    @property
+    def codebook_kind(self):
+        return self.params.codebook_kind
+
+    @property
+    def force_random_rotation(self):
+        return self.params.force_random_rotation
+
+    @property
+    def add_data_on_build(self):
+        return self.params.add_data_on_build
+
+    @property
+    def conservative_memory_allocation(self):
+        return self.params.conservative_memory_allocation
+
+
+cdef class Index:
+    """
+    IvfPq index object. This object stores the trained IvfPq index state
+    which can be used to perform nearest neighbors searches.
+    """
+
+    cdef cuvsIvfPqIndex_t index
+    cdef bool trained
+
+    def __cinit__(self):
+        self.trained = False
+        check_cuvs(cuvsIvfPqIndexCreate(&self.index))
+
+    def __dealloc__(self):
+        check_cuvs(cuvsIvfPqIndexDestroy(self.index))
+
+    @property
+    def trained(self):
+        return self.trained
+
+    def __repr__(self):
+        return "Index(type=IvfPq)"
+
+
+@auto_sync_resources
+def build(IndexParams index_params, dataset, resources=None):
+    """
+    Build the IvfPq index from the dataset for efficient search.
+
+    Parameters
+    ----------
+    index_params : IndexParams object
+    dataset : CUDA array interface compliant matrix shape (n_samples, dim)
+        Supported dtype [float, int8, uint8]
+    {resources_docstring}
+
+    Returns
+    -------
+    index: cuvs.ivf_pq.Index
+
+    Examples
+    --------
+
+    >>> import cupy as cp
+    >>> from cuvs.neighbors import ivf_pq
+    >>> n_samples = 50000
+    >>> n_features = 50
+    >>> n_queries = 1000
+    >>> k = 10
+    >>> dataset = cp.random.random_sample((n_samples, n_features),
+    ...                                   dtype=cp.float32)
+    >>> build_params = ivf_pq.IndexParams(metric="sqeuclidean")
+    >>> index = ivf_pq.build(build_params, dataset)
+    >>> distances, neighbors = ivf_pq.search(ivf_pq.SearchParams(),
+    ...                                        index, dataset,
+    ...                                        k)
+    >>> distances = cp.asarray(distances)
```

Comment:
I think we can clean this up now too and maybe follow the semantics that have been successful in cuml- if a numpy array is passed in, a numpy array should be returned, etc... 

cuml abstracts these conversations behind a "CumlArray" object. I suspect we should consider doing something similar for cuVS.